### ISSUE TITLE
formatting error when printing `model.solution` for unsolved model

### ISSUE THREAD

**Issue by pstjohn** [opencobra/cobrapy#353]
```python
In [1]: import cobra.test

In [2]: model = cobra.test.create_test_model('textbook')

In [3]: model.solution
Out[3]: ---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/Users/pstjohn/anaconda3/envs/cobra_test/lib/python3.5/site-packages/IPython/core/formatters.py in __call__(self, obj)
    670                 type_pprinters=self.type_printers,
    671                 deferred_pprinters=self.deferred_printers)
--> 672             printer.pretty(obj)
    673             printer.flush()
    674             return stream.getvalue()

/Users/pstjohn/anaconda3/envs/cobra_test/lib/python3.5/site-packages/IPython/lib/pretty.py in pretty(self, obj)
    381                             if callable(meth):
    382                                 return meth(obj, self, cycle)
--> 383             return _default_pprint(obj, self, cycle)
    384         finally:
    385             self.end_group()
```

Strange behavior on devel-2, I can dig more into this later. Just wanted to post so I don't forget.

**Comment by cdiener** [comment]
Hi when trying that I get basically the same error as in #371. So I think it's the same problem: model._timestamp_last_optimization is None and the date formatter does not accept that.

Full error:
```
TypeError                                 Traceback (most recent call last)
/home/cdiener/.local/lib/python3.5/site-packages/IPython/core/formatters.py in __call__(self, obj)
    670                 type_pprinters=self.type_printers,
    671                 deferred_pprinters=self.deferred_printers)
--> 672             printer.pretty(obj)
    673             printer.flush()
    674             return stream.getvalue()

/home/cdiener/.local/lib/python3.5/site-packages/IPython/lib/pretty.py in pretty(self, obj)
    381                             if callable(meth):
    382                                 return meth(obj, self, cycle)
--> 383             return _default_pprint(obj, self, cycle)
    384         finally:
    385             self.end_group()

/home/cdiener/.local/lib/python3.5/site-packages/IPython/lib/pretty.py in _default_pprint(obj, p, cycle)
    501     if _safe_getattr(klass, '__repr__', None) not in _baseclass_reprs:
    502         # A user-provided repr. Find newlines and replace them with p.break_()
--> 503         _repr_pprint(obj, p, cycle)
    504         return
    505     p.begin_group(1, '<')

/home/cdiener/.local/lib/python3.5/site-packages/IPython/lib/pretty.py in _repr_pprint(obj, p, cycle)
    699     """A pprint that just redirects to the normal repr function."""
    700     # Find newlines and replace them with p.break_()
--> 701     output = repr(obj)
    702     for idx,output_line in enumerate(output.splitlines()):
    703         if idx:

/home/cdiener/code/cobrapy/cobra/core/Solution.py in __repr__(self)
    104 
    105     def __repr__(self):
--> 106         if self.f is None:
    107             return "<Solution '%s' at 0x%x>" % (self.status, id(self))
    108         return "<Solution %.2f at 0x%x>" % (self.f, id(self))

/home/cdiener/code/cobrapy/cobra/core/Solution.py in f(self)
    249     @property
    250     def f(self):
--> 251         self._check_freshness()
    252         if self._f is None:
    253             return self.model.solver.objective.value

/home/cdiener/code/cobrapy/cobra/core/Solution.py in _check_freshness(self)
    239                     timestamp_formatter(self._time_stamp),
    240                     timestamp_formatter(
--> 241                         self.model._timestamp_last_optimization))
    242             )
    243 

/home/cdiener/code/cobrapy/cobra/core/Solution.py in timestamp_formatter(timestamp)
    231         if self._time_stamp != self.model._timestamp_last_optimization:
    232             def timestamp_formatter(timestamp):
--> 233                 datetime.datetime.fromtimestamp(timestamp).strftime(
    234                     "%Y-%m-%d %H:%M:%S:%f")
    235 

TypeError: a float is required
```

**Comment by pstjohn** [comment]
We should probably let the dust settle on `LazySolution` before digging too deep into this: I tried a quick fix but got bogged down in the `Solution._check_freshness()` calls. 

We should make sure the errors coming from the solution and reaction objects are predictable. For instance, the following doesn't raise any errors, and probably should:

```python
>>> model = cobra.test.create_test_model()
>>> model.reactions[0].flux
0.0
>>> model.reactions[0].reduced_cost
0.0
```

**Comment by cdiener** [comment]
Might also be worthwhile to discuss "return None vs. raise Error". The default behavior for optlang is retuirn None (unknown flux) and not raise errors, I think that is a reasonable behavior and distinguishes well from "calculated" fluxes which can never be None.

**Comment by pstjohn** [comment]
Exactly, agreed. I'm happy with None, that seems fairly unambiguous.
And we should by all means avoid/fix code like [this](https://github.com/opencobra/cobrapy/blob/devel/cobra/core/Reaction.py#L168-L179)...
I think the original intent was to raise a more informative exception, but with the newer python versions you end up seeing the entire exception chain.

**Comment by hredestig** [comment]
closed by #398 #387
